### PR TITLE
fix: simulations don't generate EVM transactions

### PR DESCRIPTION
**Pull request by Muggle-Du**
Closes: #658 

## Description

Now the evm module doesn't generate random evm txs for the application simulation.

Generally speaking, we should implement the method `WeightedOperations` for x/evm module by generating random EthereumTx messages so that we can apply random evm txs to the module when we call the `SimulateFromSeed`. When implementing `WeightedOperations`, we can retrieve previous random `EthAccount` from auth module's `AccoutKeeper`. 

By default, auth module will only generate genesis base accounts and vesting accounts in simulation. But we can pass in custom `RandomGenesisAccounts ` when initializing the application's `SimulationManager` so that we can generate genesis `EthAccount` and other account types for simulation.
```go
app.sm = module.NewSimulationManager(
		// auth.NewAppModule(appCodec, app.AccountKeeper, authsims.RandomGenesisAccounts),
		auth.NewAppModule(appCodec, app.AccountKeeper, RandomGenesisAccounts),
...
```
______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/test_helpers.go [hunk 1]
```diff
@@ -62,3 +74,40 @@ func Setup(isCheckTx bool, patchGenesis func(*EthermintApp, simapp.GenesisState)
...
+
+func RandomAccounts(r *rand.Rand, n int) []simtypes.Account {
+	accs := make([]simtypes.Account, n)
+
+	for i := 0; i < n; i++ {
+		// don't need that much entropy for simulation
+		privkeySeed := make([]byte, 15)
+		r.Read(privkeySeed)
+
+		prv := secp256k1.GenPrivKeyFromSecret(privkeySeed)
```
Comment (fedekunze) [inline app/test_helpers.go:102]: do we not have this func on the `ethsecp256k1` package?

File: app/test_helpers.go [hunk 2]
```diff
@@ -62,3 +74,40 @@ func Setup(isCheckTx bool, patchGenesis func(*EthermintApp, simapp.GenesisState)
...
+		ethPrv := &ethsecp256k1.PrivKey{}
+		ethPrv.UnmarshalAmino(prv.Bytes())
```
Comment (fedekunze) [inline app/test_helpers.go:104]: why amino?

File: x/evm/simulation/genesis.go [hunk 3]
```diff
@@ -4,16 +4,17 @@ import (
...
 	"github.com/tharsis/ethermint/x/evm/types"
 )
 
 // RandomizedGenState generates a random GenesisState for nft
 func RandomizedGenState(simState *module.SimulationState) {
-	params := types.NewParams(types.DefaultEVMDenom, true, true, types.DefaultChainConfig())
+	params := types.NewParams(sdk.DefaultBondDenom, true, true, types.DefaultChainConfig())
 	if simState.Rand.Uint32()%2 == 0 {
-		params = types.NewParams(types.DefaultEVMDenom, true, true, types.DefaultChainConfig(), 1344, 1884, 2200, 2929, 3198, 3529)
+		params = types.NewParams(sdk.DefaultBondDenom, true, true, types.DefaultChainConfig(), 1344, 1884, 2200, 2929, 3198, 3529)
```
Comment (fedekunze) [inline x/evm/simulation/genesis.go:17]: why the change?

File: x/evm/simulation/operations.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+const (
+	OpWeightMsgEthSimpleTransfer = "op_weight_msg_eth_simple_transfer"
+	OpWeightMsgEthCreateContract = "op_weight_msg_eth_create_contract"
+	OpWeightMsgEthCallContract   = "op_weight_msg_eth_call_contract"
+)
+
+const (
+	WeightMsgEthSimpleTransfer = 100
+	WeightMsgEthCreateContract = 100
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:41]: imo weight should be lower here

File: x/evm/simulation/operations.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+			SimulateEthSimpleTransfer(ak, k),
+		),
+		simulation.NewWeightedOperation(
+			weightMsgEthCreateContract,
+			SimulateEthCreateContract(ak, k),
+		),
+		// simulation.NewWeightedOperation(
+		// 	weightMsgEthCallContract,
+		// 	SimulateEthCallContract(ak, k),
+		// ),
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:98]: uncomment?

File: x/evm/simulation/operations.go [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+var (
+	maxWaitSeconds = 10
+)
+
+type SimulateContext struct {
+	context sdk.Context
+	bapp    *baseapp.BaseApp
+	rand    *rand.Rand
+	keeper  *keeper.Keeper
+}
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:58]: why public struct with unexported fields?

File: x/evm/simulation/operations.go [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+func RandomTransferableAmount(ctx *SimulateContext, address common.Address, gasLimit uint64, gasFeeCap *big.Int) (amount *big.Int, err error) {
+	balance := ctx.keeper.GetBalance(ctx.context, address)
+	feeLimit := new(big.Int).Mul(gasFeeCap, big.NewInt(int64(gasLimit)))
+	if (feeLimit.Cmp(balance)) > 0 {
+		return nil, ErrNoEnoughBalance
+	}
+	spendable := new(big.Int).Sub(balance, feeLimit)
+	if spendable.Cmp(big.NewInt(0)) == 0 {
+		amount = new(big.Int).Set(spendable)
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:250]: we should return here

File: x/evm/simulation/operations.go [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+	} else {
+		simAmount, err := simtypes.RandPositiveInt(ctx.rand, sdk.NewIntFromBigInt(spendable))
+		if err != nil {
+			return nil, err
+		}
+		amount = simAmount.BigInt()
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:256]: return

File: x/evm/simulation/operations.go [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+	signedTx, err := GetSignedTx(ctx, txBuilder, ethTx, prv)
+	if err != nil {
+		return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgEthereumTx, "can not sign ethereum tx"), nil, err
+	}
+
+	_, _, err = ctx.bapp.Deliver(txConfig.TxEncoder(), signedTx)
+
+	if err != nil {
+		return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgEthereumTx, "failed to deliver tx"), nil, err
+	}
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:197]:
```suggestion
	if err != nil {
		return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgEthereumTx, "failed to deliver tx"), nil, err
	}

```

File: x/evm/simulation/operations.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+		data := append(types.ERC20Contract.Bin, ctorArgs...)
+
+		simulateContext := &SimulateContext{ctx, bapp, r, k}
+
+		return SimulateEthTx(simulateContext, &from, contractAddr, nil, (*hexutil.Bytes)(&data), simAccount.PrivKey, nil)
+	}
+}
+
+func SimulateEthTx(ctx *SimulateContext, from, to *common.Address, amount *big.Int, data *hexutil.Bytes, prv cryptotypes.PrivKey, fops []simtypes.FutureOperation) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+	ethTx, err := CreateRandomValidEthTx(ctx, from, nil, nil, (*hexutil.Bytes)(data))
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:178]:
```suggestion
	ethTx, err := CreateRandomValidEthTx(ctx, from, nil, nil, data)
```

**Review by fedekunze — COMMENTED** [review]
First pass

**Review by fedekunze — CHANGES_REQUESTED** [review]
@Muggle-Du can you fix the linter and build errors?

**Comment by Muggle-Du** [comment]
> @Muggle-Du can you fix the linter and build errors?

Sure, fixed

File: app/test_helpers.go [hunk 11]
```diff
@@ -62,3 +74,44 @@ func Setup(isCheckTx bool, patchGenesis func(*EthermintApp, simapp.GenesisState)
 
 	return app
 }
+
+// This is used by auth module to create random genesis accounts in simulation when genesis.json not specified
+// The default auth module's RandomGenesisAccounts implementation creates only base accounts and vestings accounts
```
Comment (fedekunze) [inline app/test_helpers.go:79]:
```suggestion
// RandomGenesisAccounts is used by the auth module to create random genesis accounts in simulation when a genesis.json is not specified.
// In contrast, the default auth module's RandomGenesisAccounts implementation creates only base accounts and vestings accounts. 
```

File: app/test_helpers.go [hunk 12]
```diff
@@ -62,3 +74,44 @@ func Setup(isCheckTx bool, patchGenesis func(*EthermintApp, simapp.GenesisState)
...
+			BaseAccount: bacc,
+			CodeHash:    common.BytesToHash(emptyCodeHash).String(),
+		}
+		genesisAccs[i] = ethacc
+	}
+
+	return genesisAccs
+}
+
+// Create random accounts with ethsecp256k1.PrivKey
```
Comment (fedekunze) [inline app/test_helpers.go:96]:
```suggestion
// RandomAccounts creates random accounts with an ethsecp256k1 private key
```

File: x/evm/module.go [hunk 13]
```diff
@@ -188,5 +188,7 @@ func (AppModule) GenerateGenesisState(simState *module.SimulationState) {
 
 // WeightedOperations returns the all the evm module operations with their respective weights.
```
Comment (fedekunze) [inline x/evm/module.go:189]: we need to add the functions above too

- `RandomizedParams`
- `RegisterStoreDecoder`
- `ProposalContents`

File: x/evm/simulation/genesis.go (hunk 3 above)
Comment (fedekunze) [inline x/evm/simulation/genesis.go:17]: undo change

File: x/evm/simulation/operations.go [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+	return amount, nil
+}
+
+// TxConfig helps with pack the ethereum tx and deliver it.
+func NewTxConfig() client.TxConfig {
+	interfaceRegistry := codectypes.NewInterfaceRegistry()
+	marshaler := codec.NewProtoCodec(interfaceRegistry)
+	txConfig := tx.NewTxConfig(marshaler, tx.DefaultSignModes)
+	return txConfig
+}
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:273]: I think we can remove this and use the one from the SDK config (`sdk.GetConfig`)

**Review by fedekunze — COMMENTED** [review]
Left comments to enable the other simulations

**Comment by fedekunze** [comment]
@Muggle-Du are you planning to fix the remaining comments? cc @crypto-facs

File: x/evm/module.go (hunk 13 above)
Comment (fedekunze) [reply x/evm/module.go:191]: `RandomizedParams` is still missing

File: x/evm/simulation/operations.go [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+	simAmount, err := simtypes.RandPositiveInt(ctx.rand, sdk.NewIntFromBigInt(spendable))
+	if err != nil {
+		return nil, err
+	}
+	amount = simAmount.BigInt()
+	return amount, nil
+}
+
+// Populate the signedTx with valid values
+func GetSignedTx(ctx *simulateContext, txBuilder client.TxBuilder, msg *types.MsgEthereumTx, prv cryptotypes.PrivKey) (signedTx signing.Tx, err error) {
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:262]: update godoc, it should always start with the name of the function. The current comment is also not very descriptive

File: x/evm/simulation/operations.go [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+		Args:   args,
+		GasCap: config.DefaultGasCap,
+	})
+	if err != nil {
+		return 0, err
+	}
+	return res.Gas, nil
+}
+
+// Transferable amount is between the range [0, spendable), spendable = balance - gasFeeCap * GasLimit
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:241]: update godoc

File: x/evm/simulation/operations.go [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+// Estimate gas used by quering the keeper
+func EstimateGas(ctx *simulateContext, from, to *common.Address, data *hexutil.Bytes) (gas uint64, err error) {
+	args, err := json.Marshal(&types.TransactionArgs{To: to, From: from, Data: data})
+	if err != nil {
+		return 0, err
+	}
+
+	res, err := ctx.keeper.EstimateGas(sdk.WrapSDKContext(ctx.context), &types.EthCallRequest{
+		Args:   args,
+		GasCap: config.DefaultGasCap,
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:233]: why are we always using the default?

File: x/evm/simulation/operations.go [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+			return nil, err
+		}
+	}
+
+	ethTx = types.NewTx(ethChainID, nonce, to, amount, gasLimit, gasPrice, gasFeeCap, gasTipCap, *data, nil)
+	ethTx.From = from.String()
+	return ethTx, nil
+}
+
+// Estimate gas used by quering the keeper
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:224]: update godoc comment

File: x/evm/simulation/operations.go [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+	return simtypes.OperationMsg{}, fops, nil
+}
+
+// Populate the ethereum tx with valid random values
+func CreateRandomValidEthTx(ctx *simulateContext, from, to *common.Address, amount *big.Int, data *hexutil.Bytes) (ethTx *types.MsgEthereumTx, err error) {
+	estimateGas, err := EstimateGas(ctx, from, to, data)
+	if err != nil {
+		return nil, err
+	}
+	gasLimit := estimateGas + uint64(ctx.rand.Intn(int(sdktx.MaxGasWanted-estimateGas)))
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:204]: can you explain the rationale for this random value?

File: x/evm/simulation/operations.go [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+
+	_, _, err = ctx.bapp.Deliver(txConfig.TxEncoder(), signedTx)
+	if err != nil {
+		return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgEthereumTx, "failed to deliver tx"), nil, err
+	}
+
+	return simtypes.OperationMsg{}, fops, nil
+}
+
+// Populate the ethereum tx with valid random values
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:198]: update

File: x/evm/simulation/operations.go [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+		data = append(data, ctorArgs...)
+
+		simulateContext := &simulateContext{ctx, bapp, r, k}
+
+		return SimulateEthTx(simulateContext, &from, contractAddr, nil, (*hexutil.Bytes)(&data), simAccount.PrivKey, nil)
+	}
+}
+
+// Create valid ethereum tx and pack it as cosmos tx, and deliver it.
+func SimulateEthTx(ctx *simulateContext, from, to *common.Address, amount *big.Int, data *hexutil.Bytes, prv cryptotypes.PrivKey, fops []simtypes.FutureOperation) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:174]: update comment and split lines

File: x/evm/simulation/operations.go [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+		receipientAddr := common.BytesToAddress(tokenReceipient.Address)
+		fops[0] = simtypes.FutureOperation{
+			BlockTime: whenCall,
+			Op:        operationSimulateEthCallContract(k, &contractAddr, &receipientAddr, nil),
+		}
+		return SimulateEthTx(simulateContext, &from, nil, nil, (*hexutil.Bytes)(&data), simAccount.PrivKey, fops)
+	}
+}
+
+func operationSimulateEthCallContract(k *keeper.Keeper, contractAddr, to *common.Address, amount *big.Int) simtypes.Operation {
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:152]: this should mention that it's always calling an ERC20 contract

File: x/evm/simulation/operations.go [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+		from := common.BytesToAddress(simAccount.Address)
+		to := common.BytesToAddress(recipient.Address)
+
+		simulateContext := &simulateContext{ctx, bapp, r, k}
+
+		return SimulateEthTx(simulateContext, &from, &to, nil, (*hexutil.Bytes)(&[]byte{}), simAccount.PrivKey, nil)
+	}
+}
+
+// operationSimulateEthCallContract is the future operations of SimulateEthCreateContract to make sure valid contract call
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:115]: update comment

File: x/evm/simulation/operations.go [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+		simulation.NewWeightedOperation(
+			weightMsgEthCreateContract,
+			SimulateEthCreateContract(ak, k),
+		),
+	}
+}
+
+// Random sender, recipient and transferable amount are choosed
+// Other tx details like nonce, gasprice, gaslimit are calculated to get valid value.
+func SimulateEthSimpleTransfer(ak types.AccountKeeper, k *keeper.Keeper) simtypes.Operation {
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:95]: update comment

File: x/evm/simulation/operations.go [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+type simulateContext struct {
+	context sdk.Context
+	bapp    *baseapp.BaseApp
+	rand    *rand.Rand
+	keeper  *keeper.Keeper
+}
+
+// Two kinds of operations are generated: SimulateEthSimpleTransfer, SimulateEthCreateContract
+// Contract call operations work as the future operations of SimulateEthCreateContract
+func WeightedOperations(
```
Comment (fedekunze) [inline x/evm/simulation/operations.go:61]: update comment

File: x/evm/simulation/genesis.go [hunk 26]
```diff
@@ -4,16 +4,17 @@ import (
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/types/module"
 
 	"github.com/tharsis/ethermint/x/evm/types"
 )
 
 // RandomizedGenState generates a random GenesisState for nft
 func RandomizedGenState(simState *module.SimulationState) {
-	params := types.NewParams(types.DefaultEVMDenom, true, true, types.DefaultChainConfig())
+	params := types.NewParams(sdk.DefaultBondDenom, true, true, types.DefaultChainConfig())
```
Comment (fedekunze) [inline x/evm/simulation/genesis.go:15]:
```suggestion
	params := types.NewParams(types.DefaultEVMDenom, true, true, types.DefaultChainConfig())
```

File: x/evm/simulation/genesis.go (hunk 3 above)
Comment (fedekunze) [inline x/evm/simulation/genesis.go:17]:
```suggestion
		params = types.NewParams(types.DefaultEVMDenom, true, true, types.DefaultChainConfig(), 1344, 1884, 2200, 2929, 3198, 3529)
```

File: x/evm/simulation/genesis.go [hunk 27]
```diff
@@ -4,16 +4,17 @@ import (
 	"encoding/json"
 	"fmt"
 
+	sdk "github.com/cosmos/cosmos-sdk/types"
```
Comment (fedekunze) [inline x/evm/simulation/genesis.go:7]: instead of using the params with the sdk.DefaultBondDenom, we should replace it through an `init` function

**Review by fedekunze — CHANGES_REQUESTED** [review]
Thanks for the changes. Added more comments. Currently missing:

- `Features` changelog entry
- tests for the new `Random...` functions
- tests for the simulation package functions, specially decoder
- update Github actions to run the simulation
- define `RandomizedParams` in  `x/evm/module.go`
- update godoc comments

File: x/evm/simulation/genesis.go [hunk 28]
```diff
@@ -4,6 +4,7 @@ import (
...
```
Comment (fedekunze) [inline x/evm/simulation/genesis.go:7]:
```suggestion
```

File: x/evm/simulation/params.go [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+// ParamChanges defines the parameters that can be modified by param change proposals
+// on the simulation.
+func ParamChanges(r *rand.Rand) []simtypes.ParamChange {
+	return []simtypes.ParamChange{
+		simulation.NewSimParamChange(types.ModuleName, keyExtraEIPs,
+			func(r *rand.Rand) string {
+				return fmt.Sprintf("\"%d\"", genExtraEIPs(r))
+			},
+		),
+	}
```
Comment (fedekunze) [inline x/evm/simulation/params.go:27]: we should add param changes for the other params too

File: app/utils.go [hunk 30]
```diff
@@ -0,0 +1,175 @@
...
+		// thus simulation accounts could have positive amount of gas token.
+		bondDenom := stakingState.Params.BondDenom
+
+		evmStateBz, ok := rawState[evmtypes.ModuleName]
+		if !ok {
+			panic("staking genesis state is missing")
+		}
+
+		evmState := new(evmtypes.GenesisState)
+		err = cdc.UnmarshalJSON(evmStateBz, evmState)
```
Comment (fedekunze) [inline app/utils.go:157]: use MustUnmarshal

File: app/utils.go [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+			panic(err)
+		}
+
+		stakingStateBz, ok := rawState[stakingtypes.ModuleName]
+		if !ok {
+			panic("staking genesis state is missing")
+		}
+
+		stakingState := new(stakingtypes.GenesisState)
+		err = cdc.UnmarshalJSON(stakingStateBz, stakingState)
```
Comment (fedekunze) [inline app/utils.go:142]: ditto

File: app/utils.go [hunk 32]
```diff
@@ -0,0 +1,175 @@
...
+// AppStateFn returns the initial application state using a genesis or the simulation parameters.
+// It is a wrapper of simapp.AppStateFn to replace evm param EvmDenom with staking param BondDenom.
+func AppStateFn(cdc codec.JSONCodec, simManager *module.SimulationManager) simtypes.AppStateFn {
+	return func(r *rand.Rand, accs []simtypes.Account, config simtypes.Config,
+	) (appState json.RawMessage, simAccs []simtypes.Account, chainID string, genesisTimestamp time.Time) {
+		appStateFn := simapp.AppStateFn(cdc, simManager)
+		appState, simAccs, chainID, genesisTimestamp = appStateFn(r, accs, config)
+
+		rawState := make(map[string]json.RawMessage)
+		err := json.Unmarshal(appState, &rawState)
```
Comment (fedekunze) [inline app/utils.go:131]: we should use the cdc JSON marshaler here

File: app/utils.go [hunk 33]
```diff
@@ -0,0 +1,175 @@
...
+		}
+
+		// we should replace the EvmDenom with BondDenom
+		evmState.Params.EvmDenom = bondDenom
+
+		// change appState back
+		rawState[evmtypes.ModuleName] = cdc.MustMarshalJSON(evmState)
+
+		// replace appstate
+		appState, err = json.Marshal(rawState)
```
Comment (fedekunze) [inline app/utils.go:169]: ditto JSON marshaler from codec

**Review by fedekunze — APPROVED** [review]

File: x/evm/simulation/params.go (hunk 29 above)
Comment (fedekunze) [reply x/evm/simulation/params.go:27]: Please add as a follow-up

File: x/evm/simulation/operations.go (hunk 17 above)
Comment (fedekunze) [reply x/evm/simulation/operations.go:239]: add as follow-up

File: x/evm/simulation/operations.go (hunk 19 above)
Comment (fedekunze) [reply x/evm/simulation/operations.go:210]: add comment in follow-up

**Pull request merged**